sensu/plugins/metric-os-api.py

```diff
@@ -0,0 +1,150 @@
+        print "{0} returned {1}".format(url, e)
+        state = STATE_WARNING
+
+    return r
+
+def request(scheme, url, metric, timestamp, method='GET', retries=2, **kwargs):
+    global state
+    global metric_data
+
+    start = time.time()
+    r = None
+    try:
+        for i in range(retries + 1):
+            if i > 0:
+                time.sleep(2)
+            r = requests.request(method, url, **kwargs)
+            if r.status_code:
+                break
+    except requests.exceptions.RequestException as e:
+        print "{0} returned {1}".format(url, e)
+        state = STATE_WARNING
+
+    if r is None:
+        status = 0
+        status_code = 600
+        time_taken = -10
+        if state == STATE_OK:
+            state = STATE_WARNING
+    elif r.status_code != 200:
+        status = 0
+        status_code = r.status_code
+        time_taken = -5
+        state = STATE_CRITICAL
+    else:
+        status = 1
+        status_code = r.status_code
+        time_taken = round(time.time() - start, 3)
+
+    metric_data += '{0}.{1}.time {2} {3}\n'.format(scheme, metric, time_taken,
+                   timestamp)
+    metric_data += '{0}.{1}.code {2} {3}\n'.format(scheme, metric, status_code,
+                   timestamp)
+    metric_data += '{0}.{1}.status {2} {3}\n'.format(scheme, metric, status,
+                   timestamp)
+    return r
+
+def main():
+    timestamp = int(time.time())
+
+    parser = ArgumentParser()
+    parser.add_argument('-u', '--user', default=os.environ['OS_USERNAME'])
+    parser.add_argument('-p', '--password', default=os.environ['OS_PASSWORD'])
+    parser.add_argument('-t', '--tenant', default=os.environ['OS_TENANT_NAME'])
+    parser.add_argument('-a', '--auth-url', default=os.environ['OS_AUTH_URL'])
+    parser.add_argument('-S', '--service-type', default='compute')
+    parser.add_argument('-H', '--host')
+    parser.add_argument('-s', '--scheme', default=DEFAULT_SCHEME)
+    args = parser.parse_args()
+
+    scheme = 'sla.'+ args.scheme
+
+    url = args.auth_url + '/tokens'
+    headers = {'Accept': 'application/json', 'Content-Type': 'application/json'}
+    data = json.dumps(
+        {
+            'auth': {
+                'tenantName': args.tenant,
+                'passwordCredentials': {
+                    'username': args.user,
+                    'password': args.password
+                }
+            }
+        })
+
+
+    r = request_token(url, 'POST', 4, data=data, headers=headers)
+    token = None
+    if r:
+        access = r.json()['access']
+        token = access['token']['id']
+
+    if not token:
+        sys.exit(STATE_CRITICAL)
+
+    endpoints = {}
+    for service in access['serviceCatalog']:
+        for endpoint in service['endpoints']:
+            endpoints[service['name']] = endpoint.get('internalURL')
+
+    headers = {'Accept': 'application/json', 'X-Auth-Project-Id': args.tenant,
+               'X-Auth-Token': token}
+
+    urls = {
```

Comment:
these checks are testing the response of the API by running commands that will be affected  by  the size and/or complexity of the install.   for example if the user has a /lot/ of  networks  it will take longer to run the neutron tests ...  This may not be a problem if expectations on response times are high enough.